### ISSUE TITLE
notification alert 

### ISSUE THREAD

**Issue by d33pak-s1ngh** [tananaev/traccar#2634]
whenever any geofence or speed alert is generated it's record is stored into events table.
but i am not getting those records as alert message.
what i want is whenever any device generate geofence or speed alert user should get notified when he logins into traccer web master and email notification.
@tananaev  thanks for any kind of help.

**Comment by tananaev** [comment]
User won't be notified on login. Web notification only works if user is logged into the system, he will be shown alert immediately via web interface. Email notification should work if you configured everything correctly.

**Comment by d33pak-s1ngh** [comment]
how do i fetch user detail by its id
i have tried following api call

http://104.132.21.103:8082/api/users/24

i am passing user id with a get request.

**Comment by tananaev** [comment]
If you an admin user, you can get a full list of account:

http://104.132.21.103:8082/api/users

There is no way to get just one.

**Comment by d33pak-s1ngh** [comment]
Device {
id:integer
name:string
uniqueId:string
status:string
lastUpdate:string
positionId:integer
groupId:integer
phone:string
model:string
contact:string
category:string
geofenceIds:
attributes:attributes
} 
in device documentation while update i am getting above fields into which following are the extra fields which are not present into device table. i want to know how i can use those columns and what are the uses fo these, because i cant see those fields in web interface also.
status
contact
phone
model
category

right now i am sotring phone number and device status into attributes.

**Comment by tananaev** [comment]
Most of those fields should be in the database and in the web interface. It sounds like you have some problem with your instance of Traccar. Possibly incorrect upgrade from older version.

**Comment by d33pak-s1ngh** [comment]
how do i check my version of traccar on server? and which one is latest version?

**Comment by tananaev** [comment]
You can check your version number in the logs, but I think it's not about the version. It seems like you haven't upgraded properly.

Latest version is on the official website.

**Comment by d33pak-s1ngh** [comment]
ok thank you sir i will look into that.

**Comment by d33pak-s1ngh** [comment]
is there any way to set certain permissions for specific users like they can create device or only view device,
also in which format i have to pass date value into from and to parameter to get report,trip,route etc.
i am trying following format
2016-11-24 02:15

**Comment by tananaev** [comment]
There is a read-only option for users.

Time should be in ISO format. See official web interface for examples.

**Comment by d33pak-s1ngh** [comment]
i can see read only field into users table but in web interface there is no read only checkbox available while creating or editing user. if i set readonly directly to database nothing happens.
/home/sunsoftware/Desktop/screenshot-104.131.22.103 8082 2016-12-06 09-58-10.png

**Comment by tananaev** [comment]
Only admin can set it for other users.

Also, it seems like you are still not using latest release.

**Comment by d33pak-s1ngh** [comment]
hi @tananaev 
i have tried following **date formats** to get reports and getting errors
1.  **iso format** 2016-12-07T06:27:07.311Z
**Error**
{"details":"NullPointerException (ReportUtils:36 < Summary:72 < ReportResource:81 < ...)","message":null}

2. simple string "2016-12-07 09:15" and "2016-12-07 09:15:00 AM"
**Error:**
{"details":"Invalid format: \"2016-12-07 03:30\" is malformed at \" 03:30\" - IllegalArgumentException (... < JsonConverter:49 < ReportResource:82 < ...)","message":"Invalid format: \"2016-12-07 03:30\" is malformed at \" 03:30\""}
**Error:**
{"details":"Invalid format: \"'2016-12-07 09:15:00 AM'\" - IllegalArgumentException (... < JsonConverter:49 < ReportResource:82 < ...)","message":"Invalid format: \"'2016-12-07 09:15:00 AM'\""}
can u please tell me exact format that will be helpful thank you.

**Comment by d33pak-s1ngh** [comment]
i want summry report base on device id.
u can give me url like following or exact date format 
http://104.131.22.103:8082/api/reports/summary?deviceId=8&groupId=&from=2016-12-07T03%3A27%3A07.311Z&to=2016-12-07T06%3A27%3A07.311Z

**Comment by tananaev** [comment]
I would recommend to look at what official web interface sends.

**Comment by d33pak-s1ngh** [comment]
thanx got it from the request url from web interface.

**Comment by d33pak-s1ngh** [comment]
hello @tananaev 
i can see notification alerts on web interface as samll popup in the corner but i am not getting any email alerts. following is my settings
![notif](https://cloud.githubusercontent.com/assets/23626544/21008284/165d5160-bd67-11e6-861b-0187a95df1f5.png)

can u tell what are the settings or steps requires to get mail notification.

**Comment by tananaev** [comment]
You have to set up SMTP server.

**Comment by d33pak-s1ngh** [comment]
ok so now i have set an smtp server on my linux system which i have tested and it's able to send and recieve emails so that will be enough to recieve email notification or is there any other setting which i have to do with traccar and smtp.

**Comment by d33pak-s1ngh** [comment]
how is your notification system is working i want to implement something like yours right now i dont have any notification system what i am doing is using ajax call in every 5 seconds to check for updates. how i can use traccer notification alert with my application. i m using codeigniter php. thanks for any kind of help..

**Comment by tananaev** [comment]
You can either specify SMTP settings for the whole server in the config file, or for each user.

It's possible to forward event notification to your own server if you want.

**Comment by d33pak-s1ngh** [comment]
yes i want to do it tell me how?
also how you r getting device positions every time without any request i think u r using same notiication system.
i want to implement same feature in my server.
it would be great for my web application.

**Comment by tananaev** [comment]
How to set SMTP for each server? Through user attributes.

For updates without request, Traccar uses WebSockets.

**Comment by d33pak-s1ngh** [comment]
where i can see your websocket code for understanding?
also is there any way to store accuracy of location into position table?

**Comment by tananaev** [comment]
Client WebSocket code is in `Root.js` file.

You can store accuracy in the attributes. It's like that separate field will be added for the accuracy in future.

**Comment by d33pak-s1ngh** [comment]
so if i will jsut add an accuracy attributes into positions table it will start recording accuracy?

**Comment by d33pak-s1ngh** [comment]
for sockets can i use traccar web scoket server or i have to create everything from scrach??

**Comment by tananaev** [comment]
I don't understand the questions. Please think and do some research before asking the questions.

**Comment by d33pak-s1ngh** [comment]
sorry for that the question is if i will add an accuracy attribute into positions table then accuracy will be store into that automatically or i have to do an actual code to get accuracy from location and store it into positions manually.

**Comment by tananaev** [comment]
Obviously you would have to write the code.

**Comment by d33pak-s1ngh** [comment]
ok thanx for that.
i have a question, i have installed traccar on my server with web-interface up and running everything is fine. now i have another web-interface developed in php codeigniter and i am using api to perfrom actions which is working fine installed on same server and i have common database for both traccar and my php web-interface which is also working fine.

Now question is that how i can implement websocket to recieve  latest positions into my php web-interface??
how i can consume websocket data generated by traccar into my web interface??
i have checked your root.js file and i want to do same with my interface.
i hope the question is clear i am creating my own web interface to add some extra feature into that and i want it to work like traccar-web is working.
thank you for any  kind of help.

**Comment by tananaev** [comment]
Connect to WebSocket and wait for the data. What exactly is the issue/question?

**Comment by d33pak-s1ngh** [comment]
how do i connect to websocket? what i need to do that?

**Comment by tananaev** [comment]
Using Websockets API:

https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API

**Comment by d33pak-s1ngh** [comment]
thank u for that finally i am able to receive data from socket into my web interface.
i have another problem i can only recieve updates from socket into my web interface when i am login to traccar web once i logout from that i am getting error into my web interface
can’t establish a connection to the server. is there any soluion for that?

**Comment by tananaev** [comment]
You need to create a session.

**Comment by d33pak-s1ngh** [comment]
i am creating session at my server side in php first i call api/session then i get result which contains user details. After getting results i am storing some user details into session as following
calling session api
`$res = $client->request('POST', API_PATH . "session", array(
                        'form_params' => array(
                            'email' => $userid,
                            'password' => $password
                        ),
                        'headers' => array('Content-Type' => 'application/x-www-form-urlencoded', 'Accept' => 'application/json')
                    ));
`
storing into session
`$this->session->set_userdata(array(
                                    'userid' => $user->id,
                                    'password' => $user->password,
                                    'name' => $user->name,
                                    'email' => $user->email,
                                    'groupid' => $user->getGroupId($user->id),
                                    'isadmin' => $user->admin,
                                    'is_admin_login' => true,
                                        )
                                );`

**Comment by tananaev** [comment]
You have to set a session cookie. That's what WebSocket API uses for authentication.

**Comment by d33pak-s1ngh** [comment]
great i will try to implement that thank you for being so helpful.

**Comment by d33pak-s1ngh** [comment]
i am setting user email and password into cookie as you have done into login controller still i am getting error for connection can not be establish to server at ws://104.131.22.103:8082/api/socket.
```
set_cookie('user',$user->email,7200);
set_cookie('password',$user->password,7200);
```
following code i check into login controller
```
Ext.util.Cookies.set('user', this.lookupReference('userField').getValue(), Ext.Date.add(new Date(), Ext.Date.YEAR, 1));
Ext.util.Cookies.set('password', this.lookupReference('passwordField').getValue(), Ext.Date.add(new Date(), Ext.Date.YEAR, 1));
```

please help me.

**Comment by tananaev** [comment]
You have to set `JSESSIONID` cookie, not username and password.

**Comment by d33pak-s1ngh** [comment]
ok so i have to store response of /api/session into JSESSIONID cookie or just user name and password?

**Comment by tananaev** [comment]
`JSESSIONID` is in the cookie. I would recommend your to read about cookies and study service calls (all the parameters including header) before you ask any further questions.

**Comment by d33pak-s1ngh** [comment]
ok i read about what is JSESSIONID is a cookie sent by server into a response header wtih set-cookie parameter so when i made a api/session request to traccar server i got a cookie JSESSIONID=1ooramof41rfl14tkhrw0cxrkv;Path=/api with other paramteres into header.

now what they are saying such type of cookies are mainten by browser so when the next request is made it will add cookie into request automatically so server can know it is coming from same browser.

My problem is that i dont know how do i  use JSESSIONID so i can get websocket updates.
I dont have any session cookie because i dont know what i have to put into session cookie and how?

Request Made from my web interface to connect webscoket without login to traccar web
![screenshot from 2016-12-14 16 08 55](https://cloud.githubusercontent.com/assets/23626544/21179013/26188ca4-c218-11e6-9487-c25d320578bd.png)

Same request from my web interface with login into traccar web

![screenshot from 2016-12-14 16 10 37](https://cloud.githubusercontent.com/assets/23626544/21179045/4a53e9c4-c218-11e6-8c06-f68697a2d3b9.png)

diffrence is i have a JSESSIONID into request when i am logged in into traccar but when i am logout from traccar web my webinterface cant use JSESSIONID.
So what i need to do from here so can get it working ??
thank you for your support.

**Comment by tananaev** [comment]
I guess you have to set the cookie yourself.

**Comment by d33pak-s1ngh** [comment]
ok i tried that the problem is
for getting socket update i am sending request on following url
http://104.131.22.103:8082/api/scoket
and so the socket will use the JSESSIONID for authentication available on this path. http://104.131.22.103:8082/api/
now if i will create a cookie JSESSIONID by my self on localhost which i am doing and sending a request on following
http://104.131.22.103:8082/api/scoket 

so it will not use JSESSIONID from my localhost it will use jsessionid available on the socket server.
i hope you understand.
if i will try to create cookie from my localhost to traccar domain like following it will not allow. otherwise by doing this i would override the cookie of server an use my own which is not possible.

set_cookie('JSESSIONID',value,7200,"http://104.131.22.103:8082","/api");

**Comment by tananaev** [comment]
If it's not possible, then you should create a session from the same place you want to use websockets.

**Comment by d33pak-s1ngh** [comment]
ok apart from this are you doing anything extra to create a session. following code is taken from root.js
```
onServerReturn: function (options, success, response) {
        Ext.get('spinner').remove();
        if (success) {
            Traccar.app.setServer(Ext.decode(response.responseText));
            Ext.Ajax.request({
                scope: this,
                url: 'api/session',
                callback: this.onSessionReturn
            });
        } else {
            Traccar.app.showError(response);
        }
},
```

**Comment by d33pak-s1ngh** [comment]
```
$.ajax({
            url: "http://104.131.22.103:8082/api/session",
            dataType: "json",
            type: "POST",
            data: {
                email: "admin",
                password: "admin"
            },
            success: function(sessionResponse){
                console.log(sessionResponse);
                protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
                socket = new WebSocket(protocol + '//' + '104.131.22.103:8082' +'/api/socket');
                //openWebsocket();
            }
        });
```
Response of above code is
![screenshot from 2016-12-15 10 34 05](https://cloud.githubusercontent.com/assets/23626544/21212420/7db4d00e-c2b2-11e6-8fff-938d2fc76c52.png)

now after just getting success ful response from api/session i am calling websocket and everything is happing at same page but still websocket request is not using jsession id provided by api/session call in above response.
![screenshot from 2016-12-15 10 37 02](https://cloud.githubusercontent.com/assets/23626544/21212456/cc94bc48-c2b2-11e6-98a1-88e9cac3348a.png)

**Comment by d33pak-s1ngh** [comment]
@tananaev  @Abyss777  am i doing anything wrong for accessing websocket because i have found other people also facing this problem #2547.
as soon as i login into traccar server and make websocket call from my interface i am able to get data from socket but when i logout from traccar app and make again request for socket connection from my interface it gives error connection can not establish.
when i was login into traccar app websocket request automatically use JSESSIONID.
but without login into traccar app if i make a socket request it will not include JSESSIONID cookie into request.
and by calling api/session i am getting JSESSIONID into response header from set-cookie parameter. but browser is not using that cookie when i make request to JSESSIONID.

so is there any idea or suggetion to solve this will be helpful thanx.

**Comment by kirantpatil** [comment]
@d33pak-s1ngh, have you tried using https solution. I remember @tananaev mentioned in other issues. Please check with it once.

**Comment by kirantpatil** [comment]
https://github.com/tananaev/traccar/issues/2056 is the issue where @tananaev emphasize on using HTTPS.

**Comment by d33pak-s1ngh** [comment]
i dont have https on my server for now.

**Comment by kirantpatil** [comment]
Please refer this issue, https://github.com/tananaev/traccar/issues/2597

**Comment by d33pak-s1ngh** [comment]
@kirantpatil  idont have a domain name yet and certbot is asking for domain name but i am working with ip address so it can not generate ssl for my ip address. so getting https for now is not possible.

**Comment by meneva** [comment]
> 
> 
> Connect to WebSocket and wait for the data. What exactly is the issue/question?

Hello Tananaev, I am getting data from WebSocket connection using the sample app.js of Traccar. But it only gives me "**positions**" data not any information from "**events**" table. Is it possible to get events data along with that positions info. I just want to popup an alert on certain event.

**Comment by tananaev** [comment]
This is not the right place to ask. This is an old closed thread. If you think there is an issue, create a new thread. As for your issue though, you need to enable web notifications to receive websocket events.
